scalalib/overrides/scala/collection/mutable/ArrayBuilder.scala

```diff
@@ -0,0 +1,764 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___     Scala API                            **
+**    / __/ __// _ | / /  / _ |    (c) 2003-2013, LAMP/EPFL             **
+**  __\ \/ /__/ __ |/ /__/ __ |    http://scala-lang.org/               **
+** /____/\___/_/ |_/____/_/ | |                                         **
+**                          |/                                          **
+\*                                                                      */
+
+package scala
+package collection
+package mutable
+
+import scala.reflect.ClassTag
+import scala.runtime.BoxedUnit
+
+import scala.scalajs.js
+
+/** A builder class for arrays.
+ *
+ *  @since 2.8
+ *
+ *  @tparam T    the type of the elements for the builder.
+ */
+abstract class ArrayBuilder[T] extends Builder[T, Array[T]] with Serializable
+
+/** A companion object for array builders.
+ *
+ *  @since 2.8
+ */
+object ArrayBuilder {
+
+  /** Creates a new arraybuilder of type `T`.
+   *
+   *  @tparam T     type of the elements for the array builder, with a `ClassTag` context bound.
+   *  @return       a new empty array builder.
+   */
+  @inline
+  def make[T: ClassTag](): ArrayBuilder[T] =
+    new ArrayBuilder.generic[T](implicitly[ClassTag[T]].runtimeClass)
+
+  /** A generic ArrayBuilder optimized for Scala.js.
+   *
+   *  @tparam T              type of elements for the array builder.
+   *  @param  elementClass   runtime class of the elements in the array.
+   */
+  @inline
+  private final class generic[T](elementClass: Class[_]) extends ArrayBuilder[T] {
+
+    private val isCharArrayBuilder = classOf[Char] == elementClass
+    private val elems: js.Array[Any] = js.Array()
+
+    def +=(elem: T): this.type = {
+      elems.push(
+          if (isCharArrayBuilder) elem.asInstanceOf[Char].toInt
+          else if (elem == null) zeroOf(elementClass)
+          else elem)
+      this
+    }
+
+    def clear(): Unit =
+      elems.length = 0
+
+    def result(): Array[T] = {
+      genericArrayBuilderResult(
+          if (classOf[Unit] == elementClass) classOf[BoxedUnit]
+          else if (classOf[Null] == elementClass || classOf[Nothing] == elementClass) classOf[Object]
+          else elementClass,
```

Comment:
Consider making a tmp var with this (except if it hinders inlining). It is hard to read.